src/browser.ts

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+'use strict';
+
+import {
+  IContentsModel
+} from 'jupyter-js-services';
+
+import {
+  Message
+} from 'phosphor-messaging';
+
+import {
+  PanelLayout
+} from 'phosphor-panel';
+
+import {
+  IChangedArgs
+} from 'phosphor-properties';
+
+import {
+  Widget
+} from 'phosphor-widget';
+
+import {
+  FileButtons
+} from './buttons';
+
+import {
+  BreadCrumbs
+} from './crumbs';
+
+import {
+  DirListing
+} from './listing';
+
+import {
+  FileBrowserViewModel
+} from './viewmodel';
+
+import {
+  FILE_BROWSER_CLASS, showErrorMessage
+} from './utils';
+
+
+/**
+ * The duration of auto-refresh in ms.
+ */
+const REFRESH_DURATION = 30000;
+
+
+/**
+ * A widget which hosts a file browser.
+ *
+ * The widget uses the Jupyter Contents API to retreive contents,
+ * and presents itself as a flat list of files and directories with
+ * breadcrumbs.
+ */
+export
+class FileBrowser extends Widget {
+
+  /**
+   * Construct a new file browser.
+   *
+   * @param model - The file browser view model.
+   */
+  constructor(model: FileBrowserViewModel) {
+    super();
+    this.addClass(FILE_BROWSER_CLASS);
+    this._model = model;
+    this._model.changed.connect(this._onChanged.bind(this));
```

Comment:
`this._model.changed.connect(this._onChanged, this);`
There's no need for `.bind()` because `connect` accepts an optional `thisArg` context.